action/protocol/staking/contractstaking/cache.go

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) 2023 IoTeX Foundation
+// This source code is provided 'as is' and no warranties are given as to title or non-infringement, merchantability
+// or fitness for purpose and, to the extent permitted by law, all liability for your use of the code is disclaimed.
+// This source code is governed by Apache License 2.0 that can be found in the LICENSE file.
+
+package contractstaking
+
+import (
+	"math/big"
+
+	"github.com/iotexproject/iotex-address/address"
+	"github.com/pkg/errors"
+)
+
+type (
+	contractStakingCache struct {
+		idBucketMap           map[uint64]*bucketInfo      // map[token]bucketInfo
+		candidateBucketMap    map[string]map[uint64]bool  // map[candidate]bucket
+		idBucketTypeMap       map[uint64]*BucketType      // map[token]BucketType
+		propertyBucketTypeMap map[int64]map[uint64]uint64 // map[amount][duration]index
+		height                uint64
+		totalBucketCount      uint64 // total number of buckets including burned buckets
+	}
+)
+
+var (
+	// ErrBucketNotExist is the error when bucket does not exist
+	ErrBucketNotExist = errors.New("bucket does not exist")
+)
+
+func newContractStakingCache() *contractStakingCache {
+	cache := &contractStakingCache{
+		idBucketMap:           make(map[uint64]*bucketInfo),
+		idBucketTypeMap:       make(map[uint64]*BucketType),
+		propertyBucketTypeMap: make(map[int64]map[uint64]uint64),
+		candidateBucketMap:    make(map[string]map[uint64]bool),
+	}
+	return cache
+}
+
+func (s *contractStakingCache) GetHeight() uint64 {
+	return s.height
+}
+
+func (s *contractStakingCache) GetCandidateVotes(candidate address.Address) *big.Int {
+	votes := big.NewInt(0)
+	m, ok := s.candidateBucketMap[candidate.String()]
+	if !ok {
+		return votes
+	}
+	for id, existed := range m {
+		if !existed {
+			continue
+		}
+		bi, ok := s.idBucketMap[id]
+		if !ok {
+			continue
```

Comment:
don't consider this an error? bucket exist in candidate's map, but not exist in total map

Reply:
use `mustGetBucketInfo` instead